### PR TITLE
Implement tofile

### PR DESCRIPTION
**Pull request by jeanlaroche**
Implemented tofile. sep and format inputs are not supported and throw an error.

### CODE REVIEW

File: pythran/pythonic/include/numpy/ndarray/tofile.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+namespace numpy
+{
+
+  namespace ndarray
+  {
+    template <class T, class pS>
+    void tofile(types::ndarray<T, pS> const &expr, types::str const &file_name,
+                types::str const &sep = "", types::str const &format = "");
+
+    NUMPY_EXPR_TO_NDARRAY0_DECL(tofile);
```
Comment (serge-sans-paille) [inline pythran/pythonic/include/numpy/ndarray/tofile.hpp]: As you're doing a single pass on the data, you don't need that conversion, you could take any array expression and iterate over it!

File: pythran/pythonic/numpy/ndarray/tofile.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace numpy
+{
+
+  namespace ndarray
+  {
+    template <class T, class pS>
+    void tofile(types::ndarray<T, pS> const &expr, types::str const &file_name,
+                types::str const &sep, types::str const &format)
+    {
+      if (sep.size() != 0)
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/ndarray/tofile.hpp]: `sep.empty()` ?

File: pythran/pythonic/numpy/ndarray/tofile.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        throw types::NotImplementedError(
+            "Sep input is not implemented yet, should be left empty");
+      if (format.size() != 0)
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/ndarray/tofile.hpp]: `format.empty()`

File: pythran/pythonic/numpy/ndarray/tofile.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        throw types::NotImplementedError(
+            "Format input is not implemented yet, should be left empty");
+      std::fstream fs;
+      fs.open(file_name.c_str(), std::fstream::out | std::fstream::binary);
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/ndarray/tofile.hpp]: better: `std::ofstream`

File: pythran/pythonic/numpy/ndarray/tofile.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      if (fs.rdstate() != std::fstream::goodbit) {
+        throw types::FileNotFoundError("Could not open file " + file_name);
+      }
+      fs.write((char *)expr.buffer, sizeof(T) * expr.flat_size());
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/ndarray/tofile.hpp]: ok, it's probably easier / more efficient to use the write method than iterating over the expression. Forget about my earlier comment.

File: pythran/tables.py [hunk 6]
```diff
@@ -811,6 +811,7 @@ def update_effects(self, node):
                 Fun[[NDArray[complex, :, :, :, :]], List[complex]],
             ]
         ),
+        "tofile": ConstMethodIntr(signature=Fun[[NDArray[T0, :]], str, str],global_effects=True),
```
Comment (serge-sans-paille) [inline pythran/tables.py]: pep8: space after comma.

File: pythran/tests/test_numpy_func0.py [hunk 7]
```diff
@@ -372,6 +372,46 @@ def test_identity0(self):
...
+            os.remove(temp_name)
+
+    def test_tofile4(self):
+        temp_name = tempfile.mkstemp()[1]
+        x = numpy.random.random(1000).astype(numpy.float64)
+        try:
+            self.run_test("def np_tofile4(x,file): import numpy ; x.tofile(file); return numpy.fromfile(file)", x, temp_name, np_tofile4=[NDArray[numpy.float64,:], str])
+        finally:
+            os.remove(temp_name)
+
```
Comment (serge-sans-paille) [inline pythran/tests/test_numpy_func0.py]: The tests look nice! Maybe add an `assertRaise` to check for errors too?

File: pythran/pythonic/numpy/ndarray/tofile.hpp (hunk 4 above)
Comment (jeanlaroche) [reply pythran/pythonic/numpy/ndarray/tofile.hpp]: mmm you asked me to use fstream in fromfile!!!

**Comment by serge-sans-paille** [comment]
Looks good, can you just squash the too commits?

**Comment by jeanlaroche** [comment]
> Looks good, can you just squash the two commits?

Done

**Pull request merged**